kore/src/Kore/Strategies/Goal.hs

```diff
@@ -743,21 +747,55 @@ allPathFollowupStep claims axioms =
 removeDestination
     :: MonadCatch m
     => MonadSimplify m
+    => ProofState.ProofState goal ~ ProofState goal goal
     => FromRulePattern goal
     => ToRulePattern goal
-    => goal
-    -> Strategy.TransitionT (Rule goal) m goal
-removeDestination goal =
+    => ProofState goal goal
+    -> Strategy.TransitionT (Rule goal) m (ProofState goal goal)
+removeDestination =
+    \case
+        Goal goal ->
+            Profile.timeStrategy "Goal.RemoveDestination"
+            $ removeDestinationWorker Goal goal
+        GoalRemainder goal ->
+            Profile.timeStrategy "Goal.RemoveDestinationRemainder"
+            $ removeDestinationWorker GoalRemainder goal
+        state -> return state
+
+removeDestinationWorker
+    :: MonadCatch m
+    => MonadSimplify m
+    => ProofState.ProofState goal ~ ProofState goal goal
+    => FromRulePattern goal
+    => ToRulePattern goal
+    => (goal -> ProofState goal goal)
+    -> goal
+    -> Strategy.TransitionT (Rule goal) m (ProofState goal goal)
+removeDestinationWorker stateConstructor goal =
     errorBracket . assertEnsuresIsTop goalAsRulePattern $ do
         removal <- removalPredicate (destination right') configuration
-        let result = Conditional.andPredicate configuration removal
-        pure $ makeRuleFromPatterns goal result (Pattern.fromTermLike right')
+        if isTop removal
+            then return . stateConstructor $ goal
+            else do
+                simplifiedRemoval <-
+                    SMT.Evaluator.filterMultiOr
+                    =<< simplifyAndRemoveTopExists
+                        (Conditional.andPredicate configuration removal)
+                if not (null simplifiedRemoval)
```

Comment:
Is `simplifiedRemoval` an `OrPattern`? It would be better to use `isBottom` instead of `null`.